### DISCUSSION TITLE
Migrate from Dictionary Operations to TensorDict for Better Performance and Memory Efficiency

### DISCUSSION THREAD

**Discussion by alanlujan91** [scikit-agent/scikit-agent#120] [Ideas]
## Problem

The current codebase relies heavily on Python dictionary operations and comprehensions throughout the core computational paths, leading to several inefficiencies:

### Performance Issues
- **Excessive dictionary comprehensions**: Functions like `estimate_bellman_residual` perform multiple dictionary filtering operations per iteration:
  ```python
  # Anti-pattern: Creating new dicts on every function call
  {sym: given_vals[sym] for sym in self.arrival_variables}
  {k: v for k, v in shocks.items() if k.endswith('_0')}
  ```

- **Repeated key manipulation**: String operations on dictionary keys in hot paths:
  ```python
  # Anti-pattern: String manipulation in computational loops
  {k.replace('_0', ''): v for k, v in shocks_t.items()}
  ```

- **Memory allocations**: Dictionary merging creates temporary objects:
  ```python
  # Anti-pattern: Creates multiple intermediate dictionaries
  vals = parameters | states_t | shocks_t | controls_t
  ```

### GPU Compatibility Issues
- Dictionary operations don't leverage GPU acceleration
- Manual tensor extraction for function calls: `[vals[var] for var in signature(dr[sym]).parameters]`
- No batch processing capabilities across multiple agents/scenarios

### Code Complexity
- Scattered key filtering logic throughout the codebase
- Manual state/shock/control variable management
- Error-prone string-based variable selection

## Proposed Solution

Migrate to PyTorch's `TensorDict` for all state, shock, and parameter management. TensorDict provides:

1. **Native tensor operations** - All operations work directly on GPU tensors
2. **Batch processing** - Handle multiple agents/scenarios simultaneously  
3. **Memory efficiency** - Optimized tensor storage and operations
4. **Type safety** - Consistent tensor types throughout the pipeline

### Performance Improvements
- **~2-3x faster** Bellman residual computation (no dictionary overhead)
- **~50% memory reduction** from eliminating intermediate dictionaries
- **GPU acceleration** for all variable operations
- **Batch processing** across multiple agents/scenarios

### Code Quality
- **Eliminate 20+ dictionary comprehensions** in core functions
- **Type safety** - no more `KeyError` exceptions
- **Cleaner interfaces** - pass `TensorDict` objects instead of multiple dictionaries
- **Better debugging** - tensor shape tracking and error messages

## Implementation Plan

1. **Phase 1**: Create `AgentTensorDict` wrapper with agent-specific operations
2. **Phase 2**: Refactor `Grid` class to use TensorDict internally  
3. **Phase 3**: Update `estimate_bellman_residual` and `BellmanEquationLoss`
4. **Phase 4**: Migrate `DBlock` transition and reward functions
5. **Phase 5**: Add backward compatibility layer for existing code

## Example Transformation

**Before** (dictionary operations):
```python
# Multiple dictionary comprehensions and merging
states_dict = {sym: given_vals[sym] for sym in self.arrival_variables}
shocks_t = {k: v for k, v in shocks.items() if k.endswith('_0')}
vals = parameters | states_t | shocks_t | controls_t
reward = reward_fn(*[vals[var] for var in func_signature])
```

**After** (TensorDict operations):
```python
# Direct tensor operations, no comprehensions
states_td = input_td.select(*self.arrival_variables)
shocks_t = input_td.select_by_suffix('_0').strip_suffix('_0')
context = parameters.merge_contexts(states_td, shocks_t, controls_t)
reward = context.apply_function(reward_fn, func_signature)
```

## Success Metrics

- [ ] Zero dictionary comprehensions in hot paths (`bellman.py`, `loss.py`)
- [ ] 50%+ reduction in memory allocations during training
- [ ] Support for batch processing multiple agents
- [ ] All operations compatible with GPU tensors
- [ ] Backward compatibility maintained for existing models

This migration would significantly improve both performance and code maintainability while enabling future enhancements like multi-agent batch processing and full GPU acceleration.

**Comment by sbenthall** [comment]
I've converted this issue into a Discussion. 
- You raise some very good points.
- I don't understand where you are getting the numbers about performance improvements. What are they based on? We don't have a systematic way of benchmarking performance yet.
- I don't understand what you mean by "batch processing multiple agents"
- I also don't understand some of the examples. The tensordict example seems to refer to methods that are not in torch.
- I wonder about backwards compatibility and platform agnosticism with respect to non-torch deployments. I know Mridul 

I think the best way forward would be to break this down into actionable subtasks. What do you have in mind for `AgentTensorDict`?

The point is certainly taken that tensordicts are more performant on GPUs and maybe have some nice interfaces we could use.

**Comment by sbenthall** [comment]
A few more thoughts about this.

First, about process:

- I apologize for my initial reaction to the issue. It was an overreaction.
- I do think that given the extremely broad scope -- involving both low level implementation and high level design -- it's not a valid 'issue' in the sense that it is too broad.
- This highlights the need for *scikit-agent* to have a more formal process/workflow for _proposals_, along the lines of PEP for Python.
- Unfortunately, the project is currently at an immature phase, and is being driven by research/grant deadlines, which is in tension with steady process like proposal discussion and even pull request peer review. Ideally, those things would be decoupled, and I will work towards that.

Second, about the substance of the proposal:

- My understanding is that dictionary comprehensions are both considered good Pythonic style and are compiled to efficient execution in Python, so I'm surprised by the blanket proposal for changing them to something else. But I could easily be wrong on Python fashion. References?
- I do think it's quite reasonable that there are performance gains to be found from better GPU compatibility.
- I believe TensorDict and Dicts are both Mappable objects, and I think it's possible that at the low level it's sufficient to stick to that type. Dict is just the most easily available Mappable. A more GPU compatible Mappable data structure would be great.
- I certainly agree that the code complexity can be reduced and I think we should move that way through better functionalization or wrapping of the operations. I think that's somewhat orthogonal to the question of which data structure gets used.
- Not all of the dictionaries used in the hot paths are dictionaries of tensors. For example, some of them are dictionaries of functions: the Block dynamics, and collections of decision rules, are like this. So I think the scope of this proposal is overstated.

**Comment by sbenthall** [comment]
Linking to the other discussion about this, where we've already discussed the Grid implementation:
https://github.com/scikit-agent/scikit-agent/discussions/48
